File: MORE/LAADS/MOD/MOD08_D3.A2001.s.py
"""
This example code illustrates how to read multiple LAADS MOD08_D3 v6.1 
HDF-EOS2 Grid files in Python. This code subsets data for a specific region
and average them by season.

If you have any questions, suggestions, or comments on this example, please use
the HDF-EOS Forum (http://hdfeos.org/forums).  If you would like to see an
example of any other NASA HDF/HDF-EOS data product that is not listed in the
HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), feel free to
contact us at eoshelp@hdfgroup.org or post it at the HDF-EOS Forum
(http://hdfeos.org/forums).

Usage:  save this script and run

    $python MOD08_D3.A2001.s.py

The HDF-EOS2 files must be in your current working directory.

Tested under: Python 3.9.13 :: Miniconda
Last updated: 2023-04-17
"""
import os
import glob
import datetime

import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

from pyhdf.SD import SD, SDC

# Change this for a different data set.
DATAFIELD_NAME = "Aerosol_Optical_Depth_Land_Ocean_Mean"
# DATAFIELD_NAME = "Cloud_Top_Temperature_Mean"

i = 0
l = []

# Subset region [1].
# lon = 20 : 60 E
# lat = 0 : 30 N

latbounds = [0, 30]
lonbounds = [20, 60]

datas = None

for file in sorted(glob.glob("MOD08_D3.A200*.061.*.hdf")):
    print(file)

    hdf = SD(file, SDC.READ)

    # Read dataset.
    data_raw = hdf.select(DATAFIELD_NAME)
    data = data_raw[:, :].astype(np.double)
    xdim = hdf.select("XDim")
    lon = xdim[:].astype(np.double)

    ydim = hdf.select("YDim")
    lat = ydim[:].astype(np.double)

    # Read lat/lon & attributes only once.
    if i == 0:

        # Retrieve attributes.
        attrs = data_raw.attributes(full=1)
        lna = attrs["long_name"]
        long_name = lna[0]
        aoa = attrs["add_offset"]
        add_offset = aoa[0]
        fva = attrs["_FillValue"]
        _FillValue = fva[0]
        sfa = attrs["scale_factor"]
        scale_factor = sfa[0]
        ua = attrs["units"]
        units = ua[0]

    data[data == _FillValue] = np.nan
    data = scale_factor * (data - add_offset)
    datam = np.ma.masked_array(data, np.isnan(data))

    # Subset data / lat / lon.
    lon, lat = np.meshgrid(lon, lat)
    mask = ((lat > latbounds[0]) & (lat < latbounds[1]) &
            (lon > lonbounds[0]) & (lon < lonbounds[1]))
    
    datas = datam[mask]

    if np.isnan(datas).all():
        print("All values are NaN.")
    else:
        m = np.nanmean(datas)
        year = file[10:14]
        day = file[14:17]
        timebase = datetime.datetime(int(year), 1, 1, 0, 0, 0)
        dt = timebase + datetime.timedelta(int(day) - 1)
        mo = int(dt.month)
        print(mo)
        s = np.floor(mo / 3)
        print(s)
        if s == 4:
            s = 0
        l.append([s, m])
    i = i + 1

df = pd.DataFrame(l, columns=["Season", "Mean"])

print(df)

str_t = "MOD08_D3 2001-2002 Seasonal Average [0~30N] & [20~60E]"
t = "{0}\n{1}".format(str_t, DATAFIELD_NAME)
xl = '0=winter 1=spring 2=summer 3=fall'
plt = df.groupby("Season").mean().plot(title=t, xlabel = xl)
plt.locator_params(integer=True)

# Save image.
pngfile = "MOD08_D3.A2001.s.py.png"
fig = plt.get_figure()
fig.savefig(pngfile)

# References
# [1] https://stackoverflow.com/questions/29135885/netcdf4-extract-for-subset-of-lat-lon

File: MORE/LAADS/MOD/MOD08_D3_A2001_s.m
%
% This example code illustrates how to access and visualize the
% seasonal average of multiple LAADS MODIS MOD08_D3 L3 Grid files in MATLAB. 
%
%  If you have any questions, suggestions, comments  on this
%  example, please use the HDF-EOS Forum (http://hdfeos.org/forums). 
%
%
%   If you would like to see an  example of any other NASA HDF/HDF-EOS data
% product that is not listed in the HDF-EOS Comprehensive Examples page 
% (http://hdfeos.org/zoo), feel free to contact us at eoshelp@hdfgroup.org 
% or post it at the HDF-EOS Forum (http://hdfeos.org/forums).
%
% Usage:save this script and run (without .m at the end)
%
% $matlab -nosplash -nodesktop -r MOD08_D3_A2001_s
%
% Tested under: MATLAB R2023a
% Last updated: 2023-04-17

import matlab.io.hdfeos.*
import matlab.io.hdf4.*

% We assume that all MOD08_D3 L3 files are in the current working directory.
thepath = '.';

% Read data from a data field.
% Change this for a different data set.
DATAFIELD_NAME = 'Aerosol_Optical_Depth_Land_Ocean_Mean';
% DATAFIELD_NAME = 'Cloud_Top_Temperature_Mean';

D = dir(fullfile(thepath, 'MOD08_D3.A200*hdf'));

for k = 1:numel(D)
    FILE_NAME = fullfile(thepath, D(k).name)
    GRID_NAME='mod08';
    
    if k == 1
        % Read attributes from the data field.
        SD_id = sd.start(FILE_NAME, 'rdonly');
        sds_index = sd.nameToIndex(SD_id, DATAFIELD_NAME);
        sds_id = sd.select(SD_id, sds_index);

        % Read _FillValue from the data field.
        fillvalue_index = sd.findAttr(sds_id, '_FillValue');
        fillvalue = sd.readAttr(sds_id, fillvalue_index);

        % Read units from the data field.
        units_index = sd.findAttr(sds_id, 'units');
        units = sd.readAttr(sds_id, units_index);

        % Read scale_factor from the data field.
        scale_index = sd.findAttr(sds_id, 'scale_factor');
        scale = sd.readAttr(sds_id, scale_index);


        % Read add_offset from the data field.
        offset_index = sd.findAttr(sds_id, 'add_offset');
        offset = sd.readAttr(sds_id, offset_index);

        % Terminate access to the corresponding data set.
        sd.endAccess(sds_id);

        % Close the file.
        sd.close(SD_id);
    end
    
    % Open HDF-EOS2 file.
    file_id = gd.open(FILE_NAME, 'rdonly');
    
    % Open grid.
    grid_id = gd.attach(file_id, GRID_NAME);
    data = gd.readField(grid_id, DATAFIELD_NAME, [], [], []);
    
    % Read lat and lon data.
    lon = gd.readField(grid_id, 'XDim', [], [], []);
    lat = gd.readField(grid_id, 'YDim', [], [], []);

    % Convert the data to double type for plot.
    data=double(data);
    lon=double(lon);
    lat=double(lat);

    % Replace fill value with NaN.
    data(data==fillvalue) = NaN;

    % Filter data based on lat/lon region. (e.g., 20~60E & 0~30N)
    f_x = (lon > 20.0 & lon < 60.0);
    f_y = (lat > 0.0 & lat < 30.0);
    data = data(f_x, f_y);
                                   
    % Multiply scale and adding offset, the equation is scale *(data-offset).
    data = scale*(data-offset);
                                   
    % Detach from the Grid object.
    gd.detach(grid_id);
    gd.close(file_id);

    % Replace time value from file name.
    [filepath, name, ext] = fileparts(FILE_NAME);
    strs = split(name, ".");
    str = strs(2); % Extract day. (e.g., A2021036)

    str_year = extractBetween(str,2,5);
    year = str2double(str_year);

    str_day = extractBetween(str, 6, 8);
    day = str2double(str_day);

    str_mo = datestr(datevec(datenum(year, 1, day)), 'mm');

    s = floor(str2double(str_mo) / 3);
    if s == 4
        s = 0;
    end
    time(:) = s;
    writematrix(horzcat(time, mean(data(:), "omitmissing")), 'outs.csv', ...
                'WriteMode','append');
end

% Read filtered data to calculate average.
A = readtable('outs.csv');
B = varfun(@mean, A, 'InputVariables', 2,...
           'GroupingVariables', 1);
% Draw plot.
f = figure('Name', 'MOD08_D3 Seasonal Average', 'visible', 'off');
plot(B.Var1, B.mean_Var2);
xtickformat('%d');
xticks(B.Var1);
xlabel({'Season in 2001/03~2002/02' ; '0=winter 1=spring 2=summer 3=fall'});

% Put title.
tstring = {'MOD08_D3 Seasonal Average [0~30N] & [20~60E]';DATAFIELD_NAME};
title(tstring, 'Interpreter', 'none', 'FontSize', 16, ...
      'FontWeight','bold');

% The following fixed-size screen size will look better in PNG if
% your screen is too large.
scrsz = [1 1 800 600];
set(f, 'position', scrsz, 'PaperPositionMode', 'auto');
saveas(f, ['MOD08_D3_A2001_s.m.png']);
exit;
